File: genium/src/test/java/com/here/genium/generator/cpp/CppTypeMapperTest.java
/*
 * Copyright (C) 2016-2018 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.cpp;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.when;

import com.here.genium.model.common.Include;
import com.here.genium.model.cpp.*;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import org.franca.core.franca.FBasicTypeId;
import org.franca.core.franca.FTypeRef;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

@RunWith(JUnit4.class)
public final class CppTypeMapperTest {

  @Mock private FTypeRef francaTypeRef;

  private final CppTypeMapper typeMapper = new CppTypeMapper(null, "");

  @Before
  public void setUp() {
    MockitoAnnotations.initMocks(this);
  }

  @Test
  public void mapBoolType() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.BOOLEAN);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(cppType, CppPrimitiveTypeRef.Type.BOOL, Collections.emptyList());
  }

  @Test
  public void mapFloatType() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.FLOAT);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(cppType, CppPrimitiveTypeRef.Type.FLOAT, Collections.emptyList());
  }

  @Test
  public void mapDoubleType() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.DOUBLE);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(cppType, CppPrimitiveTypeRef.Type.DOUBLE, Collections.emptyList());
  }

  @Test
  public void mapInt8Type() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.INT8);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(
        cppType,
        CppPrimitiveTypeRef.Type.INT8,
        Collections.singletonList(CppLibraryIncludes.INT_TYPES));
  }

  @Test
  public void mapInt16Type() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.INT16);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(
        cppType,
        CppPrimitiveTypeRef.Type.INT16,
        Collections.singletonList(CppLibraryIncludes.INT_TYPES));
  }

  @Test
  public void mapInt32Type() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.INT32);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(
        cppType,
        CppPrimitiveTypeRef.Type.INT32,
        Collections.singletonList(CppLibraryIncludes.INT_TYPES));
  }

  @Test
  public void mapInt64Type() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.INT64);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(
        cppType,
        CppPrimitiveTypeRef.Type.INT64,
        Collections.singletonList(CppLibraryIncludes.INT_TYPES));
  }

  @Test
  public void mapUInt8Type() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.UINT8);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(
        cppType,
        CppPrimitiveTypeRef.Type.UINT8,
        Collections.singletonList(CppLibraryIncludes.INT_TYPES));
  }

  @Test
  public void mapUInt16Type() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.UINT16);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(
        cppType,
        CppPrimitiveTypeRef.Type.UINT16,
        Collections.singletonList(CppLibraryIncludes.INT_TYPES));
  }

  @Test
  public void mapUInt32Type() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.UINT32);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(
        cppType,
        CppPrimitiveTypeRef.Type.UINT32,
        Collections.singletonList(CppLibraryIncludes.INT_TYPES));
  }

  @Test
  public void mapUInt64Type() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.UINT64);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(
        cppType,
        CppPrimitiveTypeRef.Type.UINT64,
        Collections.singletonList(CppLibraryIncludes.INT_TYPES));
  }

  @Test
  public void mapUndefinedType() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.UNDEFINED);

    CppTypeRef cppType = typeMapper.map(francaTypeRef);

    verifyPrimitiveType(cppType, CppPrimitiveTypeRef.Type.VOID, Collections.emptyList());
  }

  @Test
  public void wrapMapType() {
    Include fooInclude = Include.createInternalInclude("bar/Foo.h");
    CppTypeRef cppTypeRef = new CppComplexTypeRef.Builder("Foo").include(fooInclude).build();

    CppTypeRef result = typeMapper.wrapMap(CppPrimitiveTypeRef.UINT32, cppTypeRef);

    assertTrue(result.includes.contains(CppLibraryIncludes.INT_TYPES));
    assertTrue(result.includes.contains(fooInclude));
    assertTrue(result.includes.contains(CppLibraryIncludes.MAP));
    assertTrue(result instanceof CppTemplateTypeRef);

    CppTemplateTypeRef cppTemplateTypeRef = (CppTemplateTypeRef) result;
    assertEquals(CppTemplateTypeRef.TemplateClass.MAP, cppTemplateTypeRef.templateClass);
    assertEquals(2, cppTemplateTypeRef.templateParameters.size());
    assertEquals(CppPrimitiveTypeRef.UINT32, cppTemplateTypeRef.templateParameters.get(0));
    assertEquals(cppTypeRef, cppTemplateTypeRef.templateParameters.get(1));
  }

  @Test
  public void wrapMapTypeWithEnumKeyType() {
    CppTypeRef cppTypeRef = new CppComplexTypeRef.Builder("Foo").refersToEnum(true).build();

    CppTypeRef result = typeMapper.wrapMap(cppTypeRef, CppPrimitiveTypeRef.VOID);

    assertTrue(result.includes.contains(CppLibraryIncludes.ENUM_HASH));
    assertTrue(result instanceof CppTemplateTypeRef);

    CppTemplateTypeRef cppTemplateTypeRef = (CppTemplateTypeRef) result;
    assertEquals(3, cppTemplateTypeRef.templateParameters.size());
    assertEquals(typeMapper.getEnumHashType(), cppTemplateTypeRef.templateParameters.get(2));
  }

  private void verifyPrimitiveType(
      CppTypeRef actual, CppPrimitiveTypeRef.Type expectedType, List<Include> includes) {
    assertTrue(actual instanceof CppPrimitiveTypeRef);
    CppPrimitiveTypeRef primitiveType = (CppPrimitiveTypeRef) actual;
    assertEquals(expectedType, primitiveType.type);
    assertEquals(new HashSet<>(includes), primitiveType.includes);
  }
}

File: genium/src/test/java/com/here/genium/generator/cpp/CppTypeMapperTest.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.cpp

import com.here.genium.model.common.Include
import com.here.genium.model.cpp.CppComplexTypeRef
import com.here.genium.model.cpp.CppPrimitiveTypeRef
import com.here.genium.model.cpp.CppTemplateTypeRef
import com.here.genium.model.cpp.CppTypeDefRef
import com.here.genium.model.lime.LimeArray
import com.here.genium.model.lime.LimeAttributeType
import com.here.genium.model.lime.LimeAttributes
import com.here.genium.model.lime.LimeBasicType.TypeId
import com.here.genium.model.lime.LimeBasicTypeRef
import com.here.genium.model.lime.LimeContainer
import com.here.genium.model.lime.LimeDirectTypeRef
import com.here.genium.model.lime.LimeEnumeration
import com.here.genium.model.lime.LimeMap
import com.here.genium.model.lime.LimePath.Companion.EMPTY_PATH
import com.here.genium.model.lime.LimeSet
import com.here.genium.model.lime.LimeStruct
import com.here.genium.model.lime.LimeTypeDef
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class CppTypeMapperTest {
    @MockK private lateinit var includeResolver: CppIncludeResolver
    @MockK private lateinit var nameResolver: CppNameResolver

    private lateinit var typeMapper: CppTypeMapper

    @Before
    fun setUp() {
        MockKAnnotations.init(this, relaxed = true)
        typeMapper = CppTypeMapper(nameResolver, includeResolver, emptyList())

        every { nameResolver.getFullyQualifiedName(any()) } returns "Foo"
        every {
            includeResolver.resolveIncludes(any())
        } returns listOf(Include.createInternalInclude("Bar"))
    }

    @Test
    fun mapTypeDefRef() {
        val limeType = LimeTypeDef(EMPTY_PATH, typeRef = LimeBasicTypeRef.DOUBLE)
        val limeTypeRef = LimeDirectTypeRef(limeType)

        val result = typeMapper.mapType(limeTypeRef)

        assertTrue(result is CppTypeDefRef)
        assertEquals("Foo", result.fullyQualifiedName)
        assertEquals("Bar", result.includes.first().fileName)
    }

    @Test
    fun mapStructRef() {
        val limeType = LimeStruct(EMPTY_PATH)
        val limeTypeRef = LimeDirectTypeRef(limeType)

        val result = typeMapper.mapType(limeTypeRef)

        assertTrue(result is CppComplexTypeRef)
        assertEquals("Foo", result.fullyQualifiedName)
        assertEquals("Bar", result.includes.first().fileName)
    }

    @Test
    fun mapEnumRef() {
        val limeType = LimeEnumeration(EMPTY_PATH)
        val limeTypeRef = LimeDirectTypeRef(limeType)

        val result = typeMapper.mapType(limeTypeRef)

        assertTrue(result is CppComplexTypeRef)
        assertEquals("Foo", result.fullyQualifiedName)
        assertEquals("Bar", result.includes.first().fileName)
        assertEquals(typeMapper.enumHashType, result.hashType)
        assertTrue(result.refersToValueType)
    }

    @Test
    fun mapArrayRef() {
        val limeType = LimeArray(LimeBasicTypeRef.DOUBLE)
        val limeTypeRef = LimeDirectTypeRef(limeType)

        val result = typeMapper.mapType(limeTypeRef)

        assertTrue(result is CppTemplateTypeRef)
        assertEquals(
            CppTemplateTypeRef.TemplateClass.VECTOR,
            (result as CppTemplateTypeRef).templateClass
        )
        assertEquals(CppPrimitiveTypeRef.DOUBLE, result.templateParameters.first())
    }

    @Test
    fun mapMapRef() {
        val limeType = LimeMap(LimeBasicTypeRef(TypeId.STRING), LimeBasicTypeRef.DOUBLE)
        val limeTypeRef = LimeDirectTypeRef(limeType)

        val result = typeMapper.mapType(limeTypeRef)

        assertTrue(result is CppTemplateTypeRef)
        assertEquals(
            CppTemplateTypeRef.TemplateClass.MAP,
            (result as CppTemplateTypeRef).templateClass
        )
        assertEquals(CppTypeMapper.STRING_TYPE, result.templateParameters.first())
        assertEquals(CppPrimitiveTypeRef.DOUBLE, result.templateParameters.last())
    }

    @Test
    fun mapMapEnumRef() {
        val limeEnumType = LimeEnumeration(EMPTY_PATH)
        val limeEnumTypeRef = LimeDirectTypeRef(limeEnumType)
        val limeType = LimeMap(limeEnumTypeRef, LimeBasicTypeRef.DOUBLE)
        val limeTypeRef = LimeDirectTypeRef(limeType)

        val result = typeMapper.mapType(limeTypeRef)

        assertTrue(result is CppTemplateTypeRef)
        assertEquals(
            typeMapper.enumHashType,
            (result as CppTemplateTypeRef).templateParameters.last()
        )
    }

    @Test
    fun mapInstanceRef() {
        val limeType = LimeContainer(EMPTY_PATH, type = LimeContainer.ContainerType.TYPE_COLLECTION)
        val limeTypeRef = LimeDirectTypeRef(limeType)

        val result = typeMapper.mapType(limeTypeRef)

        assertTrue(result is CppTemplateTypeRef)
        assertEquals(
            CppTemplateTypeRef.TemplateClass.SHARED_POINTER,
            (result as CppTemplateTypeRef).templateClass
        )

        val instanceType = result.templateParameters.first()
        assertTrue(instanceType is CppComplexTypeRef)
        assertEquals("Foo", instanceType.fullyQualifiedName)
        assertEquals("Bar", instanceType.includes.first().fileName)
        assertTrue((instanceType as CppComplexTypeRef).needsForwardDeclaration)
    }

    @Test
    fun mapExternalInstanceRef() {
        val limeType = LimeContainer(
            EMPTY_PATH,
            type = LimeContainer.ContainerType.TYPE_COLLECTION,
            attributes = LimeAttributes.Builder()
                .addAttribute(LimeAttributeType.EXTERNAL_TYPE)
                .build()
        )
        val limeTypeRef = LimeDirectTypeRef(limeType)

        val result = typeMapper.mapType(limeTypeRef)

        val instanceType = (result as CppTemplateTypeRef).templateParameters.first()
        assertTrue(instanceType is CppComplexTypeRef)
        assertEquals("Foo", instanceType.fullyQualifiedName)
        assertEquals("Bar", instanceType.includes.first().fileName)
        assertFalse((instanceType as CppComplexTypeRef).needsForwardDeclaration)
    }

    @Test
    fun mapSetTypeRef() {
        val limeType = LimeSet(LimeBasicTypeRef.DOUBLE)
        val limeTypeRef = LimeDirectTypeRef(limeType)

        val result = typeMapper.mapType(limeTypeRef)

        assertTrue(result is CppTemplateTypeRef)
        assertEquals(
            CppTemplateTypeRef.TemplateClass.SET,
            (result as CppTemplateTypeRef).templateClass
        )
        assertEquals(CppPrimitiveTypeRef.DOUBLE, result.templateParameters.first())
    }

    @Test
    fun mapEnumSetRef() {
        val limeEnumType = LimeEnumeration(EMPTY_PATH)
        val limeEnumTypeRef = LimeDirectTypeRef(limeEnumType)
        val limeType = LimeSet(limeEnumTypeRef)
        val limeTypeRef = LimeDirectTypeRef(limeType)

        val result = typeMapper.mapType(limeTypeRef)

        assertTrue(result is CppTemplateTypeRef)
        assertEquals(
            typeMapper.enumHashType,
            (result as CppTemplateTypeRef).templateParameters.last()
        )
    }

    @Test
    fun mapOptionalType() {
        val limeTypeRef = LimeBasicTypeRef.DOUBLE.asNullable()

        val result = typeMapper.mapType(limeTypeRef)

        assertEquals(
            CppTemplateTypeRef.TemplateClass.OPTIONAL,
            (result as CppTemplateTypeRef).templateClass
        )
    }
}
